### DISCUSSION TITLE
Popover hint on icon-only buttons

### DISCUSSION THREAD

**Discussion by janhassel** [carbon-design-system/carbon#14593] [Ideas]
# The problem

Using icon-only buttons that will trigger a menu of additional options is fairly common in UI designs across the industry. The depiction of three dots (aligned horizontally or vertically) is essentially an established standard by now that conveys this information. However, in many cases this kebap/meatball icon isn't appropriate and a more descriptive icon should be used when similar options are grouped. For this scenario, the Carbon `OverflowMenu` offers a `renderIcon` prop to change the trigger icon. Without any additional affordance though, it's hard for users to know whether the button will now already perform an action or if they'll receive a list of options (in form of a menu).

This has led to inconsistent experiences where some teams are using no additional indicator and some teams are using a chevron-down next to the icon which can result to unfavorable spacing and visual relationship issues (also it's not supported by the Carbon components).

In the design guidance, Carbon describes a [Text Toolbar pattern](https://carbondesignsystem.com/patterns/text-toolbar-pattern/) (which isn't available in code afaik). Here, a small caret is used in the bottom right corner to distinguish the icon-only menu buttons from the action performing buttons.

![Text Toolbar pattern: Menus](https://carbondesignsystem.com/static/ebedbaf418ee923f1ac77af2ba17af4c/92c65/text-toolbar-menus.png)

# Existing patterns

## Figma
The Figma toolbar uses a chevron. However, this is used inconsistently as the very first item (Figma logo) is a menu button (entire button is the menu trigger) and the rest are actually combo buttons where only the chevron is the menu trigger.
<img width="412" alt="image" src="https://github.com/carbon-design-system/carbon/assets/28265588/9e31a8cd-30d6-4712-8a27-4197ee43f856">

## macOS Notes
This uses a chevron and works as a menu button.
<img width="185" alt="image" src="https://github.com/carbon-design-system/carbon/assets/28265588/2d1a8c6e-4ca5-44fc-a2a6-50d15c67732b">

## GitHub header actions
Also using a chevron and a menu button pattern. Here, the buttons are clearly visually separated though.
<img width="231" alt="image" src="https://github.com/carbon-design-system/carbon/assets/28265588/281cc4a9-cce9-4abf-9669-ab91d3cd7605">

## Spotify
No visual affordance between action-performing buttons and menu buttons.
<img width="338" alt="image" src="https://github.com/carbon-design-system/carbon/assets/28265588/b5f9baeb-e6da-478a-b57a-a54a1244c26c">

## Photoshop toolbar
Using a caret in the bottom right corner. The interaction pattern is different though: Single click: tool is selected, long or right-click: menu appears.
<img width="39" alt="image" src="https://github.com/carbon-design-system/carbon/assets/28265588/53f10103-f6bd-49a8-a19e-4758cbef8bd3">

## Blender
Using a chevron and menu button pattern. The chevrons are placed very closely to the icon and further visual grouping is established though background and border styling.
<img width="82" alt="image" src="https://github.com/carbon-design-system/carbon/assets/28265588/7b9b5338-37d7-41c2-aedd-da93a68b7f7c">

# Approaches

## 1. Visual indicator
Define a visual indicator that should be used when the trigger icon isn't clearly communicating the menu button pattern.
In code, `OverflowMenu` could render this indicator by default when `renderIcon` is used. There should probably be a way to opt-out of this though with another prop (sth. like `props.noPopoverHint`). Alternatively make it explicit opt-in (`props.showPopoverHint`).

## 2. IconMenuButton
Create an `IconMenuButton` variant based on the `MenuButton` component which will break out of the square boundaries of an IconButton. The padding could be controlled to avoid the spacing and visual relationship issues described above.
If spacing isn't an option, this component could render an inline border similar to the divider in `ButtonSet`

## 3. Design guidance only
Without offering code, clearly guide designers on this topic and how to prevent above mentioned issues.*

_*full disclosure, I'm not a big fan of this one as I believe the current inconsistencies we experience in products using Carbon stems from the fact that there is no coded solution to this problem._

---

cc @aagonzales @mbgower @thyhmdo

**Comment by mbgower** [comment]
The [popover variants](https://carbondesignsystem.com/components/popover/usage/#variants), particular the No Tip, which requires a visual cue on the trigger, are very relevant to this discussion
> Popovers without a tip are typically used when the trigger button has a visually defined down state

 It's also worth looking at the disclosure pattern info on [visual guidance](https://carbondesignsystem.com/patterns/disclosures-pattern/#visual-guidance)

**Comment by aagonzales** [comment]
Some examples I've run across in the wild

<img width="920" alt="image" src="https://github.com/carbon-design-system/carbon/assets/11670886/e365ba7f-69ec-4f6f-bd42-8c18a6cf2e31">

**Comment by thyhmdo** [comment]
More other existing components with the `caret-down` icon



----
### Revit

<img width="700" alt="Screenshot 2023-09-08 at 8 00 36 AM" src="https://github.com/carbon-design-system/carbon/assets/77697351/5c0af944-ffa5-4e0e-a5fd-40fd93614f78">

![image](https://github.com/carbon-design-system/carbon/assets/77697351/0988f342-3422-44ac-bc08-4f389be1389d)

---
### Sketchup

![image](https://github.com/carbon-design-system/carbon/assets/77697351/cd613d21-f482-4fa1-98b8-934bf1544423)
![image](https://github.com/carbon-design-system/carbon/assets/77697351/c65be131-7108-405d-afd7-0635c4a3dc38)


---
### AutoCad

![image](https://github.com/carbon-design-system/carbon/assets/77697351/929c7239-179b-4506-bff5-08ff38686cf1)

**Comment by AmandaFurukawa** [comment]
This is a product example that exists in Cognos Analytics - note that this isn't using the carbon toolbar, it's custom.
![image](https://github.com/carbon-design-system/carbon/assets/119629745/0385ac0b-b170-4c0b-a2ff-f34e98638def)
![image](https://github.com/carbon-design-system/carbon/assets/119629745/b05e0f2e-1deb-4de2-95a5-54172ee2d643)

**Comment by AmandaFurukawa** [comment]
Planning Analytics has the same pattern 
<img width="1690" alt="image" src="https://github.com/carbon-design-system/carbon/assets/119629745/e78877b2-d152-475b-857e-03821713c45c">

<img width="1689" alt="image" src="https://github.com/carbon-design-system/carbon/assets/119629745/b4fc0ef4-ef20-4369-aabe-a6b3b55ccfe9">

**Comment by thyhmdo** [comment]
The [article](https://www.nngroup.com/articles/split-buttons/)  pointed out that this example (bottom right corner) may not be recommended, as there is no visual signifier for that unique affordance

![image](https://github.com/carbon-design-system/carbon/assets/77697351/7e7fb64d-8513-44a2-99e5-3282d53bbc4f)

**Comment by mbgower** [comment]
> The article pointed out that this example (bottom right corner) may not be recommended, as there is no visual signifier for that unique affordance

Note that they're critiquing that as a split button design, wherein the user need to get to the other options with a press and hold. So, it's not that they're criticizing the arrow, so much as criticizing the split button pattern. It's a little frustrating they do not link to any specific guidance on button menus (as opposed to split buttons) in that article.

I do not think Jan is looking for a split button. Just a button menu with a visual dropdown indicator. Most of the visual examples folks have added have been for split buttons.
